### ISSUE TITLE
need stable client ip hashing in case of both fail-over and recovery

### ISSUE THREAD

**Issue by AmitBawer** [yyyar/gobetween#191]
Currently release 0.6.1 allows using iphash1 for client-ip hashing load balancing strategy.
yet according to documentation it is guaranteed that active back-end traffic will remain redirected as is in case another back-end fails. When the failed back-end recovers it is possible that active back-ends traffic will get reshuffled as in the original iphash method. 
We need a client-ip hashing method which guarantees that the active traffic is kept stable in both cases.

**Comment by AmitBawer** [comment]
btw if current iphash1 is stable both cases when discovery kind="static" than it should suffice. However it is not mentioned in GB documentation/code.

**Comment by nickdoikov** [comment]
@illarion  could you please add more description to Iphash1/iphash ?

**Comment by AmitBawer** [comment]
@nickdoikov - is it known what is the behavior for iphash1 in a case of static backends list where no new servers are discovered but  some can fail and recover?

**Comment by illarion** [comment]
@AmitBawer yes it is known and it's really simple: in case of failure of some backend, its clients get rebalanced to other backends. Other clients are not rebalanced. If the affected backend recovers, only its clients get rebalanced back.

**Comment by AmitBawer** [comment]
Thank you @illarion , it sure sounds good!

**Comment by yyyar** [comment]
Thanks for your contribution, @AmitBawer 
Sticky iphash implementation seems something very nice to have in gobetween.
We'll work on "sticky" version as new iphash2 balance method.
I think it will accept some parameters like expiration time and max sticky table size, like in haproxy.

**Comment by AmitBawer** [comment]
@yyyar Thank you indeed, looking forward for such functionality.

**Comment by AmitBawer** [comment]
@shantanugadgil , i believe that if no rebalance never occurs after recovery we might end up theoretically in situation where all clients are forever redirected to the ever-healthy backend and the recovered one would remain idle. Unless you have meant something else.

**Comment by cunningwasabi** [comment]
> Thanks for your contribution, @AmitBawer
> Sticky iphash implementation seems something very nice to have in gobetween.
> We'll work on "sticky" version as new iphash2 balance method.
> I think it will accept some parameters like expiration time and max sticky table size, like in haproxy.

I wonder if sticky is actually just a reasonable option for any of the balance methods.
ie, you might like 'weight' to be sticky/not sticky, same for iphash, priority, etc etc.
I recently submitted a pull request for a 'priority' balance mechanism, and am extending that mechanism to include another called 'stickyPriority'.
if you wrapped "stickyXXX" to just do the "sticky" component, then call the normal balance function "XXX" for new sessions, you'd probably have a nice extensible framework moving forward?

**Comment by cunningwasabi** [comment]
so, I wrote 'sticky priority' and submitted it, see #209
